### ISSUE TITLE
定义损失函数和优化器

### ISSUE THREAD

**Issue by renxianxu** [renxianxu/MathStudy-Machine-Learning#4]
import torch
import torch.nn as nn
import torch.optim as optim
import torch.nn.functional as F
from torch.utils.data import DataLoader
from torchvision import datasets, transforms

# 1. 定义损失计算函数，计算每个小批量（mini-batch）的损失
def loss_batch(loss_func, xb, yb, yb_h, opt=None):
    # 计算损失
    loss = loss_func(yb_h, yb)
    # 计算性能指标（例如准确度）
    metric_b = metrics_batch(yb, yb_h)
    
    # 如果存在优化器，则执行反向传播
    if opt is not None:
        loss.backward()  # 反向传播
        opt.step()  # 更新权重
        opt.zero_grad()  # 清除梯度
    
    return loss.item(), metric_b  # 返回损失值和性能指标

# 2. 定义计算每个小批量准确度的函数
def metrics_batch(target, output):
    # 选择输出中概率最大的类别作为预测值
    pred = output.argmax(dim=1, keepdim=True)
    # 比较预测结果和真实标签，计算正确的数量
    corrects = pred.eq(target.view_as(pred)).sum().item()
    return corrects

# 3. 定义计算整个数据集损失和指标的函数
def loss_epoch(model, loss_func, dataset_dl, opt=None):
    loss = 0.0
    metric = 0.0
    len_data = len(dataset_dl.dataset)  # 获取数据集的大小
    
    # 遍历数据集，计算每个小批量的损失和指标
    for xb, yb in dataset_dl:
        xb = xb.type(torch.float).to(device)  # 将输入数据移动到设备
        yb = yb.to(device)  # 将标签移动到设备
        
        # 获取模型的输出
        yb_h = model(xb)
        
        # 计算损失和指标
        loss_b, metric_b = loss_batch(loss_func, xb, yb, yb_h, opt)
        loss += loss_b
        if metric_b is not None:
            metric += metric_b
    
    # 返回整个数据集的平均损失和平均准确度
    loss /= len_data
    metric /= len_data
    return loss, metric

# 4. 定义训练和验证过程的函数
def train_val(epochs, model, loss_func, opt, train_dl, val_dl):
    for epoch in range(epochs):
        # 训练模式
        model.train()
        # 计算训练损失和训练指标
        train_loss, train_metric = loss_epoch(model, loss_func, train_dl, opt)
        
        # 验证模式
        model.eval()  
        with torch.no_grad():  # 在验证过程中不需要计算梯度
            val_loss, val_metric = loss_epoch(model, loss_func, val_dl)
        
        # 计算验证准确率
        accuracy = 100 * val_metric / len(val_dl.dataset)
        
        # 输出当前epoch的训练和验证信息
        print(f"epoch: {epoch + 1}, train loss: {train_loss:.6f}, val loss: {val_loss:.6f}, accuracy: {accuracy:.2f}%")

# 5. 定义神经网络模型
class Net(nn.Module):
    def __init__(self):
        super(Net, self).__init__()
        self.conv1 = nn.Conv2d(1, 20, 5, 1)
        self.conv2 = nn.Conv2d(20, 50, 5, 1)
        self.fc1 = nn.Linear(4 * 4 * 50, 500)
        self.fc2 = nn.Linear(500, 10)

    def forward(self, x):
        x = F.relu(self.conv1(x))
        x = F.max_pool2d(x, 2, 2)
        x = F.relu(self.conv2(x))
        x = F.max_pool2d(x, 2, 2)
        x = x.view(-1, 4 * 4 * 50)
        x = F.relu(self.fc1(x))
        x = self.fc2(x)
        return F.log_softmax(x, dim=1)

# 6. 初始化模型、损失函数、优化器
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")  # 使用GPU或CPU
model = Net().to(device)  # 将模型移到设备（GPU或CPU）

# 使用交叉熵损失函数，适合分类任务
loss_func = nn.CrossEntropyLoss()
# 使用Adam优化器
opt = optim.Adam(model.parameters(), lr=0.001)

# 7. 加载训练集和验证集数据加载器
transform = transforms.Compose([transforms.ToTensor(), transforms.Normalize((0.5,), (0.5,))])

# 加载MNIST数据集
train_data = datasets.MNIST(root='./data', train=True, download=True, transform=transform)
val_data = datasets.MNIST(root='./data', train=False, download=True, transform=transform)

# 使用DataLoader包装训练集和验证集
train_dl = DataLoader(train_data, batch_size=64, shuffle=True)
val_dl = DataLoader(val_data, batch_size=64, shuffle=False)

# 8. 训练模型
num_epochs = 5
train_val(num_epochs, model, loss_func, opt, train_dl, val_dl)
